xwiki-platform-core/xwiki-platform-vfs/xwiki-platform-vfs-api/src/main/java/org/xwiki/vfs/VfsResourceReference.java

```diff
@@ -95,24 +105,77 @@ public VfsResourceReference(URI uri, String pathSegments)
 
     /**
      * @param fullURI the full opaque URI containing both the reference to the archive and the path to the entry inside
-     *        it, e.g. {@code attach:space.page@attachment/path/to/file}. Note that this constructor requires that any
-     *        "/" character inside the reference to the archive be URL-encoded
+     * it, e.g. {@code attach:space.page@attachment/path/to/file}. Note that this constructor requires that the
+     * full URL to be URL-encoded.
+     * @deprecated Since 12.3, this constructor shouldn't be used anymore, in particular not for internal references
+     * such as {@code attach:space.page@attachment/path/to/file}, {@link #VfsResourceReference(String)}
+     * should be used instead
      */
+    @Deprecated
     public VfsResourceReference(URI fullURI)
     {
         // Find the first "/" and consider that everything after is the path
-        this(URI.create(StringUtils.substringBefore(fullURI.toString(), RESOURCE_PATH_SEPARATOR)),
-            StringUtils.substringAfter(fullURI.toString(), RESOURCE_PATH_SEPARATOR));
+        this(URI.create(retrieveRootReference(fullURI.toString())), retrievePathSegment(fullURI.toString()));
+    }
+
+    /**
+     * @param fullReference the full opaque reference containing both the reference to the archive and the path to the
+     * entry inside it, e.g. {@code attach:space.page@attachment/path/to/file}.
+     * @since 12.3
+     */
+    @Unstable
+    public VfsResourceReference(String fullReference)
+    {
+        super(XWIKI_REFERENCE, EntityResourceAction.fromString(""));
+
+        setType(TYPE);
+        String referenceWithScheme = retrieveRootReference(fullReference);
+        this.scheme = StringUtils.substringBefore(referenceWithScheme, SCHEME_SEPARATOR);
+        this.reference = StringUtils.substringAfter(referenceWithScheme, SCHEME_SEPARATOR);
+        this.pathSegments = Arrays.asList(
+            StringUtils.split(retrievePathSegment(fullReference), RESOURCE_PATH_SEPARATOR));
+    }
+
+    private static String retrieveRootReference(String fullReference)
+    {
+        return StringUtils.substringBefore(fullReference, RESOURCE_PATH_SEPARATOR);
+    }
+
+    private static String retrievePathSegment(String fullReference)
+    {
+        return StringUtils.substringAfter(fullReference, RESOURCE_PATH_SEPARATOR);
     }
 
     /**
      * @return the URI to the VFS (e.g. {@code attach:space.page@file.zip}, {@code http://server/path/to/zip})
+     * @deprecated Since 12.3 this method shouldn't be used anymore, in favor of {@link #getReference}.
      */
+    @Deprecated
     public URI getURI()
     {
         return this.uri;
```

Comment:
Should this try to create a URI with the old logic when it's null for retro compatibility ?

Reply:
Could be safer indeed.